django/contrib/auth/forms.py

```diff
@@ -178,9 +178,12 @@ def __init__(self, request=None, *args, **kwargs):
         self.user_cache = None
         super(AuthenticationForm, self).__init__(*args, **kwargs)
 
-        # Set the label for the "username" field.
+        # Set the label and username length for the "username" field.
         UserModel = get_user_model()
         self.username_field = UserModel._meta.get_field(UserModel.USERNAME_FIELD)
+        self.fields['username'].widget.attrs['maxlength'] = UserModel._meta.get_field(
+            UserModel.USERNAME_FIELD
+        ).max_length
```

Comment:
`max_length` can be `None` for non-`CharField` instances of `Field`. You might want to assign the attribute only if it is not `None`.